Realm/ObjectServerTests/SwiftObjectServerTests.swift

```diff
@@ -472,6 +472,90 @@ class SwiftObjectServerTests: SwiftSyncTestCase {
+            try prepareClientReset(#function, user)
+
+            var configuration = user.configuration(partitionValue: #function)
+            configuration.objectTypes = [SwiftPerson.self]
+
+            let ex = self.expectation(description: "get client reset error")
+            let syncManager = self.app.syncManager
+            syncManager.errorHandler = { error, session in
+                guard let error = error as? SyncError else {
+                    XCTFail("Bad error type: \(error)")
+                    return
+                }
+                XCTAssertEqual(error.code, .clientResetError)
+                XCTAssertEqual(session?.state, .inactive)
+                XCTAssertEqual(session?.connectionState, .disconnected)
+                XCTAssertEqual(session?.parentUser()?.id, user.id)
+                ex.fulfill()
+            }
+
+            try autoreleasepool {
+                let realm = try Realm(configuration: configuration)
+                wait(for: [ex], timeout: 60.0)
+                // The locally created object should still be present as we didn't
+                // actually handle the client reset
+                XCTAssertEqual(realm.objects(SwiftPerson.self).count, 1)
+                XCTAssertEqual(realm.objects(SwiftPerson.self)[0].firstName, "John")
+            }
+            syncManager.waitForSessionTermination()
+            resetSyncManager()
+        }
+
+        let user = try logInUser(for: creds)
+        var configuration = user.configuration(partitionValue: #function)
+        configuration.objectTypes = [SwiftPerson.self]
+
+        try autoreleasepool {
+            let realm = try Realm(configuration: configuration)
+            waitForDownloads(for: realm)
+            // After reopening, the old Realm file should have been moved aside
+            // and we should now have the data from the server
+            XCTAssertEqual(realm.objects(SwiftPerson.self).count, 1)
+            XCTAssertEqual(realm.objects(SwiftPerson.self)[0].firstName, "Paul")
+        }
+    }
+
+    func testClientResetManualWithoutLiveRealmInstance() throws {
+        let creds = basicCredentials()
+        let user = try logInUser(for: creds)
+        try prepareClientReset(#function, user)
+
+        var configuration = user.configuration(partitionValue: #function)
+        configuration.objectTypes = [SwiftPerson.self]
+
+        let ex = self.expectation(description: "get client reset error")
+        let syncManager = self.app.syncManager
+        syncManager.errorHandler = { error, session in
+            guard let error = error as? SyncError else {
+                XCTFail("Bad error type: \(error)")
+                return
+            }
+            XCTAssertEqual(error.code, .clientResetError)
+            XCTAssertEqual(session?.state, .inactive)
+            XCTAssertEqual(session?.connectionState, .disconnected)
+            XCTAssertEqual(session?.parentUser()?.id, user.id)
```

Comment:
So I understand this test correctly:
If the change in RLMSyncConfiguration.mm weren't there then the `session` in 540-542 would be in an invalid state/not exist because it would have been deallocated because there were no strong references.

Reply:
Yes, without the fix the third check on 542 failed because the `id` would be `nil`. It was technically a "valid" state, but just not a very useful one.